### PR TITLE
Introducing SequenceDecoder for Avro

### PR DESCRIPTION
**Pull request by thomasabraham**
team-aruha/issues/1056

### CODE REVIEW

File: core-common/src/main/java/org/zalando/nakadi/repository/kafka/AvroDeserializerWithSequenceDecoder.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            final GenericRecord metadata = metadataSequenceDecoder.read(envelop.getMetadata());
+
+            metadata.put("occurred_at", new DateTime(
+                    (long) metadata.get("occurred_at"), DateTimeZone.UTC).toString());
+            metadata.put("received_at", new DateTime(
+                    (long) metadata.get("received_at"), DateTimeZone.UTC).toString());
+
+            final GenericRecord event = eventSequenceDecoder.read(envelop.getPayload());
+
+            final StringBuilder sEvent = new StringBuilder(event.toString());
```
Comment (a1exsh) [inline core-common/src/main/java/org/zalando/nakadi/repository/kafka/AvroDeserializerWithSequenceDecoder.java:33]: Does the `toString()` call produce JSON?

File: core-common/src/main/java/org/zalando/nakadi/repository/kafka/AvroDeserializerWithSequenceDecoder.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    private final SequenceDecoder metadataSequenceDecoder;
+    private final SequenceDecoder eventSequenceDecoder;
+
+    public AvroDeserializerWithSequenceDecoder(final AvroSchema schemas) {
+        this.metadataSequenceDecoder = new SequenceDecoder(schemas.getMetadataSchema());
+        this.eventSequenceDecoder = new SequenceDecoder(schemas.getNakadiAccessLogSchema());
+    }
+
+    public byte[] deserializeAvro(final EnvelopeHolder envelop) throws RuntimeException {
```
Comment (a1exsh) [inline core-common/src/main/java/org/zalando/nakadi/repository/kafka/AvroDeserializerWithSequenceDecoder.java:22]:
```suggestion
    public byte[] deserializeAvro(final EnvelopeHolder envelop) {
```
It's not a checked exception, so no need to declare it.

File: core-common/src/main/java/org/zalando/nakadi/repository/kafka/SequenceDecoder.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            inputStreams.add(inputStream);
+        }
+    }
+
+    public SequenceDecoder(final Schema schema) {
+        this.datumReader = new GenericDatumReader<>(schema);
+        //Adding an empty stream so that Building SequenceInputStream won't close it immediately.
+        inputStreamEnumeration.add(new ByteArrayInputStream(new byte[0]));
+        this.binaryDecoder = DecoderFactory.get()
+                .directBinaryDecoder(new SequenceInputStream(inputStreamEnumeration), null);
```
Comment (a1exsh) [inline core-common/src/main/java/org/zalando/nakadi/repository/kafka/SequenceDecoder.java:46]: Wow!  Java truly has utilities for all and everything one can only imagine %)

**Comment by a1exsh** [comment]
:+1:

File: core-common/src/main/java/org/zalando/nakadi/repository/kafka/AvroDeserializerWithSequenceDecoder.java (hunk 1 above)
Comment (a1exsh) [reply core-common/src/main/java/org/zalando/nakadi/repository/kafka/AvroDeserializerWithSequenceDecoder.java:33]: Given that the tests are green, I assume it does.

**Comment by thomasabraham** [comment]
:+1:

**Pull request merged**